java-frontend/src/main/java/org/sonar/java/se/symbolicvalues/RelationalSymbolicValue.java

```diff
@@ -209,42 +198,202 @@ RelationalSymbolicValue inverse() {
   }
 
   private boolean checkRelation(BooleanConstraint booleanConstraint, ProgramState programState) {
-    RelationState relationState = binaryRelation().resolveState(programState.getKnownRelations());
+    RelationState relationState = resolveState(programState);
     return !relationState.rejects(booleanConstraint);
   }
 
-  private List<RelationalSymbolicValue> transitiveRelations(ProgramState programState) {
-    BinaryRelation relation = binaryRelation();
-    return programState.getKnownRelations().stream()
-      .map(r -> r.deduceTransitiveOrSimplified(relation))
-      .filter(Objects::nonNull)
-      .map(RelationalSymbolicValue::binaryRelationToSymbolicValue)
-      .collect(Collectors.toList());
+  RelationState resolveState(ProgramState programState) {
+    if (hasSameOperand()) {
+      return relationStateForSameOperand();
+    }
+
+    Set<RelationalSymbolicValue> allRelations = knownRelations(programState).collect(Collectors.toCollection(HashSet::new));
```

Comment:
No, order is not relevant here, we are testing all relations with current relation, if any of them resolves `this` we return the result